### PR TITLE
ota-blog-part3: Deploying OTA Community Edition

### PR DESCRIPTION
**Pull request by doanac**
Signed-off-by: Andy Doan <andy@opensourcefoundries.com>

### CODE REVIEW

**Comment by doanac** [comment]
Sorry guys - this is a bit of a monster.

File: src/content/blog/ota-part-3.md [hunk 1]
```diff
@@ -0,0 +1,234 @@
++++
+title = "Deploying OTA Community Edition"
+date = "2018-06-27"
+tags = ["ota", "open source"]
+categories = ["FOTA"]
+banner = "img/banners/ota.png"
++++
+
+Continuing with the OTA blog series [part one]({{< ref "ota-part-1.md" >}}) and [part two]({{< ref "ota-part-2.md" >}}), this article shows you how to deploy OTA Community Edition inside Google's Kubernetes Engine (GKE). After completion of these instructions, you'll have an OTA server available on the internet with a single QEMU device registered to it.
```
Comment (mbolivar) [inline src/content/blog/ota-part-3.md]: Just a note: when this is all over, it'd be nice to go back and update each entry `i` by adding a header paragraph saying:

```
This is entry i in a series of n blog posts about over-the-air updates in the Linux microPlatform. List of entries:

- link to entry 1, with title
- link to entry 2, with title
- ..
- link to entry n, with title
```

File: src/content/blog/ota-part-3.md [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+### Requirements
+
+You'll need a few things in place to be able to deploy this to GKE:
+
+ * An account with the [Google Cloud Platform](https://console.cloud.google.com) (GCP).
+ * A Project defined in your GCP account.
+ * Git and Docker to run the deployment scripts.
+
+Its also highly recommended that you are able to add DNS records to a domain you own. Otherwise, you'll have to edit /etc/hosts to access the services.
```
Comment (mbolivar) [inline src/content/blog/ota-part-3.md]: s/Its also/It's also/

File: src/content/blog/ota-part-3.md [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+### Overview
+
+OTA Community Edition deploys several [micro-services]({{< ref "ota-part-2.md" >}}) to Kubernetes. However, it doesn't come with any built-in security. The approach taken by this article is to expose the unsafe services via a single nginx reverse-proxy. This will make it obvious what services need to be locked down. Part four of the blog series will describe an approach for securing the reverse-proxy. When the deployment is completed, you'll have a system with these caveats:
+
+ * **SECURE** - connection of devices to the OTA gateway
+ * **SECURE** - TUF keys will be offline
+ * **INSECURE** - Anyone can push to the TreeHub
```
Comment (mbolivar) [inline src/content/blog/ota-part-3.md]: I'm confused about if these "insecure" things are just for what we'll have at the end of part 3 and will have instructions for securing in part 4, or if they'll stay insecure forever (and how bad that is).

File: src/content/blog/ota-part-3.md [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+~~~
+
+### Configure OTA Community Edition
+
+OTA Community Edition provides a way to override their default configuration.
+You'll need to create a `config/local.yaml` as documented here::
+~~~
+  # config/local.yaml
+  #
+  # I prefer to not use Ingress controllers in GKE for a few reasons:
```
Comment (mbolivar) [inline src/content/blog/ota-part-3.md]: If you're going to use "I", I think it'd be better to add yourself as an author in the config block at the top of file:

```
author = "Andy Doan"
```

File: src/content/blog/ota-part-3.md [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+  # 1) They are still in beta.
+  # 2) They cost a little more than just doing a Service of type "loadBalancer".
+  # 3) When you deploy this securely, you'll need your own nginx reverse-proxy
+  #    that can handle authenticate/authorization.
```
Comment (mbolivar) [inline src/content/blog/ota-part-3.md]: "authentication" would be consistent with "authorization"

File: src/content/blog/ota-part-3.md [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+  create_ingress: false
+
+  # We aren't using an Ingress, but templates/services/app.tmpl.yaml still
+  # uses this for building up the DNS names of the services it uses. This
+  # field must be set to the domain you host your service under. eg:
```
Comment (mbolivar) [inline src/content/blog/ota-part-3.md]: "under, e.g.:"

File: src/content/blog/ota-part-3.md [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+  # ingress_dns_name: foundriez.io
+  ingress_dns_name: <DNS_NAME>
+
+  # Use the 0.4 ota-tuf containers. They don't require "vault" which makes
```
Comment (mbolivar) [inline src/content/blog/ota-part-3.md]: Should this be "at least version 0.4 of", or is it exactly 0.4 that matters?

File: src/content/blog/ota-part-3.md [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+  # 1) They are still in beta.
+  # 2) They cost a little more than just doing a Service of type "loadBalancer".
+  # 3) When you deploy this securely, you'll need your own nginx reverse-proxy
+  #    that can handle authenticate/authorization.
+  create_ingress: false
+
+  # We aren't using an Ingress, but templates/services/app.tmpl.yaml still
+  # uses this for building up the DNS names of the services it uses. This
+  # field must be set to the domain you host your service under. eg:
+  # ingress_dns_name: foundriez.io
```
Comment (mbolivar) [inline src/content/blog/ota-part-3.md]: How about an RFC 2606 style "foundries.example" or "example.com"?

File: src/content/blog/ota-part-3.md [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+
+  zookeeper_mem: 500Mi
+  zookeeper_disk: 80Gi
+~~~
+
+Now you can generate the files required by the OTA Gateway:
+~~~
+  # For this example we'll deploy to "foundriez.io". The DNS_NAME here
+  # *must* must match the ingress_dns_name in config/local.yaml.
+  ./contrib/gke/make SERVER_NAME="ota-ce.foundriez.io" DNS_NAME="foundriez.io" new-server
```
Comment (mbolivar) [inline src/content/blog/ota-part-3.md]: RFC 2606 style here too? (and 2 lines above)

File: src/content/blog/ota-part-3.md [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+~~~
+  # Start the "infrastructure". This is MySQL, ZooKeeper, and Kafka. (Takes about 3 minutes)
+  ./contrib/gke/make start-infra
+
+  # Do a quick sanity check to verify all 3 StatefulSets are running.
+  ./contrib/gke/kubectl get statefulsets
+
+  # Now deploy OTA Community Edition. (Takes 2-3 minutes)
+  # DNS_NAME and SERVER_NAME *must* match what you used above for "new-server".
+  ./contrib/gke/make SERVER_NAME="ota-ce.foundriez.io" DNS_NAME="foundriez.io" start-services
```
Comment (mbolivar) [inline src/content/blog/ota-part-3.md]: Here too

File: src/content/blog/ota-part-3.md [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+
+As a sanity check, you can validate everything is working with:
+~~~
+  curl http://tuf-reposerver.<DNS_NAME>/api/v1/user_repo/targets.json
+  curl http://treehub.<DNS_NAME>/api/v3/config
+~~~
+
+### Secure the TUF Keys
+
+OTA Community Edition is running and exposed to the internet. The nginx reverse-proxy has no built-in security and the TUF private keys are available to **anyone**. This can be fixed even though the reverse-proxy has no authentication/authorization logic. In fact, its a great example of how TUF protects you from the leaking of a private key:
```
Comment (mbolivar) [inline src/content/blog/ota-part-3.md]: s/its a great/it's a great/

**Comment by doanac** [comment]
I just pushed a change to address @mbolivar's comments. The next thing I'm going to do is pull break out 2 large chunks of the blog and make them well documented shell scripts that the user can run. It should make things easier to read and deploy

File: src/content/blog/ota-part-3.md (hunk 3 above)
Comment (rsalveti) [reply src/content/blog/ota-part-3.md]: Guess maybe just describing the insecure pieces instead?

**Comment by doanac** [comment]
i cherry-picked to the new repo also

**Pull request merged**